docs/RFCS/20220711_external_connections.md

```diff
@@ -0,0 +1,407 @@
+// the external resource represented by an External Connection.
+type ConnectionDetails interface {
+  // ConnectionType returns the type of the External Connection.
+  ConnectionType() connectionpb.ConnectionType
+  // ConnectionProto prepares the ConnectionDetails for serialization.
+  ConnectionProto() *connectionpb.ExternalConnectionDetails
+  // Dial establishes a connection to the external resource.
+  Dial() Connection 
+  // DisplayString returns the string representation of the
+  // external resource with all sensitive information redacted.
+  DisplayString() string
+}
+
+// Connection is a marker interface that is implemented by objects that
+// support interaction with the external resource.  
+type Connection interface {}
+```
+
+The `ConnectionProto` method returns the concrete `connectionpb.ExternalConnectionDetails` that can
+be serialized and stored in the `connection_details` columns of the system table.
+The `ConnectionType` method returns the connection type that the External Connection belongs to.
+
+To prevent misuse of the underlying endpoint URI and configuration information, we will expose an
+opaque `Dial` method. This method will be responsible for establishing a connection to the external
+resource and returning a handle to interact with it. The `Connection` interface is a marker
+interface that will be implemented by the `cloud.ExternalStorage`, `cloud.KMS`,
+and `changefeedccl.Sink` interfaces described in the section below. The handle will therefore
+support all the methods that are used by bulk operations, and changefeeds to interact with external
+connections today.
+
+The `DisplayString` method returns a string representation of the external connection with all
+secrets redacted. This string will be used by observability tools that display information about the
+external resource. The scheme-specific implementations have control over how to display their
+endpoint URI and configurations as well as what information they consider privileged.
+
+### Accessing an External Connection
+
+Database operations such as backup, restore, import, export, and changefeeds interact with external
+resources via user-inputted URIs as part of the SQL query. Depending on the database operation and
+the scheme of the URI, the parsing, validation, and creation of a handle to the external connection
+are managed by different packages in CockroachDB. Each package recognizes a set of URI schemes and
+in turn returns one of the following interfaces to interact with the external connection:
+
+`cloud.ExternalStorage`: Used by bulk operations to read and write files.
+
+`cloud.KMS`: Used by bulk operations to encrypt and decrypt data using a Key Management Service.
+
+`changefeedccl.Sink`: Used by changefeeds to stream changes to an external connection.
+
+This RFC proposes introducing a new URI scheme that can be used with all bulk operations and
+changefeeds. The URI will be of the form `external://<external-connection-name>`.
```

Comment:
another idea: we could just use the name, bare, with no scheme/hostname split. This would end up meaning that when we try to parse it as a URI, it all ends up in the scheme, so we'd check if the parsed scheme is "http", or "s3" etc as we do now, but then the `default` cause would be to try to look it up the value of parsed.Scheme as a name for a named connection lookup.

Thus the statement would look something like `BACKUP INTO 'mybackups';` to use the `mybackups` named connection, or `s3://foo/bar` to use an explicit uri. 

This does suggest that we'd either want to reserve the names of the schemes used for explicit URIs to prevent using them as connection names, or I suppose just offer the explicit external:// qualifier where we use the Host component as an option if someone really, really wants to name their connection "s3".

Reply:
I'd rather have less magic.

Reply:
I'm generally in the "less magic" camp as well and am by no means going to bat for this, but my reason for thinking it might be nice is that it really emphasizes that if you start typing out a URI, you're really switching to manual control vs the preferred "just tell me the name" flow.